Userland/Applications/Maps/MapsWidget.cpp

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2023, Bastiaan van der Plaat <bastiaan.v.d.plaat@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "MapsWidget.h"
+#include <AK/URL.h>
+#include <LibDesktop/Launcher.h>
+#include <LibGfx/ImageFormats/ImageDecoder.h>
+#include <LibProtocol/Request.h>
+
+MapsWidget::MapsWidget(Options const& options)
+    : m_request_client(Protocol::RequestClient::try_create().release_value_but_fixme_should_propagate_errors())
+    , m_tile_layer_url(options.tile_layer_url)
+    , m_center(options.center)
+    , m_zoom(options.zoom)
+    , m_attribution_enabled(options.attribution_enabled)
+    , m_attribution_text(options.attribution_text)
+    , m_attribution_url(options.attribution_url)
+{
+}
+
+void MapsWidget::mousedown_event(GUI::MouseEvent& event)
+{
+    if (event.buttons() != GUI::MouseButton::Primary)
+        return;
+
+    // Handle attribution text click
+    if (m_attribution_enabled) {
+        GUI::Painter painter(*this);
+        float attribution_width = 8 + painter.font().width(m_attribution_text) + 8;
+        float attribution_height = 8 + painter.font().pixel_size() + 8;
+        if (
+            static_cast<float>(event.x()) > static_cast<float>(width()) - attribution_width && static_cast<float>(event.y()) > static_cast<float>(height()) - attribution_height) {
+            Desktop::Launcher::open(m_attribution_url);
+            return;
+        }
+    }
+
+    // Start map tiles dragging
+    m_dragging = true;
+    m_last_mouse_x = event.x();
+    m_last_mouse_y = event.y();
+    set_override_cursor(Gfx::StandardCursor::Drag);
+}
+
+void MapsWidget::mousemove_event(GUI::MouseEvent& event)
+{
+    if (!m_dragging)
+        return;
+
+    double delta_x = event.x() - m_last_mouse_x;
+    double delta_y = event.y() - m_last_mouse_y;
+
+    // Adjust map center latitude and longitude by mouse delta
+    double tile_size = TILE_SIZE * pow(2, m_zoom);
+    m_center.latitude -= (-delta_y / tile_size) * 180;
+    m_center.latitude = min(max(m_center.latitude, -MAX_LATITUDE), MAX_LATITUDE);
```

Comment:
The latitude movement feels a bit off. Dragging vertically doesn't quite follow the cursor. (Dragging from the top to the bottom of the screen, the map moves about 2/3 as far as it should.) Longitude is perfect though.

Reply:
You are right the math was wrong. I have reworked and simplified the math in the new code and dragging works correct now.